### PR TITLE
feat: enforce json schema for config json

### LINKED ISSUE #543: Enforce json schema for config.json
As a user I want to know that I've configured my experiment correctly and know the possible values per setting
So the experiment will fail fast and not in the middle of a long run.

```[tasklist]
### Tasks
- [ ] add [jsonschema](https://pypi.org/project/jsonschema/) package
- [ ] create json schema for all possible values for config.json
- [ ] validate json schema and throw an error upon running any step
- [ ] validate json schema upon pre-commit for sample and ci configuration
```

Technical Notes:

- https://www.linkedin.com/pulse/mastering-json-schema-validation-python-developers-guide-singh-ebffc/

### PR DESCRIPTION
**Pull request by liammoat**
This pull request introduces JSON Schema validation to improve application configuration.

### Using `$schema

Json Scheme can be configured to use a local scheme:

```json
{
    "$schema": "./config.schema.json",
    "index_name_prefix": "idx",
    ...
}
```

Or a remote URI:

```json
{
    "$schema": "https://raw.githubusercontent.com/microsoft/rag-experiment-accelerator/limoat/enforce-json-schema-for-configjson/config.schema.json",
    "index_name_prefix": "idx",
    ...
}
```

I propose, the `config.sample.json` uses the remote URI by default.

### Notable Changes

The core of this PR is the `config.schema.json` file. In addition, the following configuration files have been updated to use the `$schema`:
  * [`config_validator.py`](diffhunk://#diff-e5d8693712c4538f5337e1cc8a7e7ceb1c151dd477bec75612cfae4bac5a738cR39): Added a validator for config, using `jsonschema`.
  * [`config.py`](diffhunk://#diff-6278f8d119fc93cb5a20073790e143b578a19eddab0d9ff678baebb203387f69R81-R83): Updated `config.py` to call `validate_json_with_schema` and raise an error if validation fails
  * [`.github/workflows/config.json`](diffhunk://#diff-17c14d82d805e2aea15d9dee2aafd1c94a29b3df330b2baac7b9ea9d9888de5eR2): Added a reference to the local schema to validate the configuration.
  * [`config.sample.json`](diffhunk://#diff-d455f18fa91da0c6b9d4e13aec2494ec66519ad2ee036318bb85441f173df4bfR2): Added a reference to the remote schema to validate the configuration.
* [`.pre-commit-config.yaml`](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R14-R21): Added `pre-commit` hook to validate config files on commit.

### Output

Here is an example of the error output when running `01_index.py` when a schema violation:
image

This error highlights:
- An error summary: "'my-transformer' is not one of ['azure', 'sentence-transformer']"
- A snapshot of the schema, including a property description: "Type of the embedding model. `azure` for Azure OpenAI models and `sentence-transformer` for HuggingFace"
- A location for the violation in the config file: "On instance['embedding_models'][0]['type']"

Resolves #543,  #345

### CODE REVIEW

**Comment by julia-meshcheryakova** [comment]
- [ ] Do not merge until prerelease in merged into development. #556

**Review by julia-meshcheryakova — APPROVED** [review]
LGTM (fixed conflicts after configs refactoring)

File: config.sample.json
Comment (liammoat) [reply config.sample.json:2]: As a sample, which users of the project will reference, I propose we keep this as a user content URL. Once it's merged into main, we can update to the main path 👍

**Review by julia-meshcheryakova — APPROVED** [review]
Looks good now

**Pull request merged**
